### ISSUE TITLE
rubocop-daemon-wrapper: netcat doesn't exit forever on Ubuntu

### ISSUE THREAD

**Issue by fohte** [fohte/rubocop-daemon#4]
On Ubuntu, netcat to talk to the `rubocop-daemon` server doesn't exit forever.

netcat is also used by `rubocop-daemon-wrapper`:

https://github.com/fohte/rubocop-daemon/blob/1f14eb779708fc7e837bf144801470cd28fa6a6f/bin/rubocop-daemon-wrapper#L42

So, `rubocop-daemon-wrapper` doesn't exit forever because of this issue.

I think this issue is caused by the difference from some other versions of netcat in its treatment of EOF on the input.

To give an example, the Ubuntu version of netcat doesn't shutdown the network socket after EOF on the input by default.
But this behavior can change by passing `-N` or `-q` option to shutdown as expected.

```
-N      shutdown(2) the network socket after EOF on the input.  Some servers require this to finish their work.
-q seconds
             after EOF on stdin, wait the specified number of seconds and then quit. If seconds is negative, wait forever (default).  Specifying a non-negative seconds implies -N.
```

My workaround on Ubuntu is:

```patch
diff --git a/bin/rubocop-daemon-wrapper b/bin/rubocop-daemon-wrapper
index 058955d..9a36237 100755
--- a/bin/rubocop-daemon-wrapper
+++ b/bin/rubocop-daemon-wrapper
@@ -39,7 +39,7 @@ run_rubocop_command() {
   PORT="$(cat "$PORT_PATH")"
   COMMAND="$TOKEN $PROJECT_ROOT exec $@"
   rm -f "$STATUS_PATH" # Clear the previous status
-  if echo -e "$COMMAND${STDIN_CONTENT}" | nc localhost "$PORT"; then
+  if echo -e "$COMMAND${STDIN_CONTENT}" | nc -N localhost "$PORT"; then
     if [ -f "$STATUS_PATH" ]; then
       exit "$(cat $STATUS_PATH)"
     else
```

This workaround works well, but `-N` option is not supported on macOS.

Therefore, we should wrap netcat command or rewrite the `rubocop-daemon-wrapper` in Go, Rust or and so on to support each platform.

### Environment

* OS: Ubuntu 18.04.1 LTS
* `rubocop-daemon` version: v0.3.0
* `nc` version: OpenBSD netcat (Debian patchlevel 1.187-1ubuntu0.1)

**Comment by ndbroadbent** [comment]
Hi @fohte, sorry about that! I've opened a PR to fix that issue and detect the OS: #5 

But I agree, we should use Go or Rust for this (and it would be a little bit faster as well.)

**Comment by ndbroadbent** [comment]
Whoa, I just realized that `rubocop-daemon` could be integrated directly with the [`vscode-ruby`](https://github.com/rubyide/vscode-ruby) extension. That would make it even faster, because then it wouldn't need to start any client processes. It can just send the TCP packets directly. 

Very excited to start working on that!

**Comment by fohte** [comment]
Many thanks to @ndbroadbent, I've merged #5 to fix this issue!

And directly sending of TCP packets with the vscode-ruby extension seems like a clean solution! :heart_eyes:
